WalletWasabi.Fluent.Generators/UIContextGenerator.cs

```diff
@@ -0,0 +1,332 @@
+		var namespaceName = classSymbol.ContainingNamespace.ToDisplayString();
+
+		var ctors =
+			classDeclaration.ChildNodes()
+							.OfType<ConstructorDeclarationSyntax>()
+							.ToList();
+
+		foreach (var ctor in ctors)
+		{
+			if (!classDeclaration.GetUiContextReferences(semanticModel).Any())
+			{
+				if (!classDeclaration.IsAbstractClass(semanticModel) && ctor.IsPublic())
+				{
+					yield return ctor;
+				}
+			}
+			// if constructor already has a UIContext parameter, leave it be. Don't generate a new ctor and use the current one for FluentNavigation.
+			else if (ctor.ParameterList.Parameters.Any(p => p.Type.IsUiContextType(semanticModel)))
+			{
+				// it must be public though
+				if (ctor.IsPublic())
+				{
+					yield return ctor;
+				}
+			}
+			else
+			{
+				var ctorArgs =
+					  ctor.ParameterList.Parameters
+						  .Select(x => x.Identifier.ValueText)
+						  .ToList();
+
+				var hasCtorArgs = ctorArgs.Any();
+				var ctorArgsString = string.Join(",", ctorArgs);
+				var ctorString =
+					hasCtorArgs
+					? $": this({ctorArgsString})"
+					: "";
+
+				var parameterUsings =
+					ctor.ParameterList.Parameters
+									  .Where(p => p.Type is not null)
+									  .Select(p => semanticModel.GetTypeInfo(p.Type))
+									  .Where(t => t.Type is not null)
+									  .Select(t => $"using {t.Type.ContainingNamespace.ToDisplayString()};")
+									  .ToList();
+
+				var parametersString = "UiContext uiContext";
+
+				var uiContextParameter =
+					SyntaxFactory.Parameter(SyntaxFactory.Identifier("uiContext")
+														 .WithLeadingTrivia(SyntaxFactory.Space))
+								 .WithType(SyntaxFactory.ParseTypeName("UiContext"));
+
+				if (hasCtorArgs)
+				{
+					parametersString += ", ";
+				}
+
+				parametersString =
+					ctor.ParameterList.Parameters.Insert(0, uiContextParameter).ToFullString();
+
+				var usings = string.Join(Environment.NewLine, parameterUsings.Distinct().OrderBy(x => x));
+
+				var code =
+	$$"""
+{{usings}}
+using WalletWasabi.Fluent.Models.UI;
+
+namespace {{namespaceName}};
+
+partial class {{className}}
+{
+    public {{className}}({{parametersString}}){{ctorString}}
+    {
+	    UiContext = uiContext;
+    }
+}
+""";
+
+				var sourceText = SourceText.From(code, Encoding.UTF8);
+				context.AddSource(fileName, sourceText);
+
+				var tree = SyntaxFactory.ParseSyntaxTree(sourceText);
+
+				var newConstructor =
+					tree.GetRoot()
+						.DescendantNodes()
+						.OfType<ConstructorDeclarationSyntax>()
+						.FirstOrDefault();
+
+				yield return newConstructor;
+			}
+		}
+	}
+
+	private static void GenerateFluentNavigation(GeneratorExecutionContext context, IEnumerable<ConstructorDeclarationSyntax> ctors)
+	{
+		var compilation = context.Compilation;
+		var namespaces = new List<string>();
+		var methods = new List<string>();
+
+		var newSyntaxTrees =
+			ctors.Select(x => x.SyntaxTree)
+				 .Where(x => !compilation.ContainsSyntaxTree(x))
+				 .ToList();
```

Comment:
The same here as well: `ToList` can be replaced by `ToArray`, and not sure about the additional line break after the assignement.